File: Linear_Control_Course/Chapter24/Lesson2/Chapter24_Lesson2.java
public final class Complex {
    public final double re, im;
    public Complex(double re, double im) { this.re = re; this.im = im; }
    public Complex add(Complex o) { return new Complex(re + o.re, im + o.im); }
    public Complex sub(Complex o) { return new Complex(re - o.re, im - o.im); }
    public Complex mul(Complex o) {
        return new Complex(re * o.re - im * o.im, re * o.im + im * o.re);
    }
    public Complex div(Complex o) {
        double den = o.re * o.re + o.im * o.im;
        return new Complex((re * o.re + im * o.im) / den,
                           (im * o.re - re * o.im) / den);
    }
    public double abs() { return Math.hypot(re, im); }
    public static final Complex ONE = new Complex(1.0, 0.0);
}

// Example: evaluate S(jw), T(jw) on a grid
public class RobustPerfCheck {

    static Complex Pjw(double w, double zeta, double wn) {
        Complex jw = new Complex(0.0, w);
        Complex denom = jw.mul(jw)
            .add(new Complex(2.0 * zeta * wn, 0.0).mul(jw))
            .add(new Complex(wn * wn, 0.0));
        return Complex.ONE.div(denom);
    }

    static Complex Cjw(double w, double K, double z) {
        Complex jw = new Complex(0.0, w);
        return new Complex(K, 0.0)
            .mul(jw.add(new Complex(z, 0.0)))
            .div(jw);
    }

    public static void main(String[] args) {
        double zeta = 0.7, wn = 10.0;
        double K = 20.0, z = 2.0;

        double maxS = 0.0, maxT = 0.0;
        for (int i = 0; i <= 500; ++i) {
            double frac = i / 500.0;
            double w = Math.pow(10.0, -2.0 + 5.0 * frac); // 1e-2..1e3

            Complex P = Pjw(w, zeta, wn);
            Complex C = Cjw(w, K, z);
            Complex L = C.mul(P);
            Complex denom = Complex.ONE.add(L);
            Complex S = Complex.ONE.div(denom);
            Complex T = L.div(denom);

            maxS = Math.max(maxS, S.abs());
            maxT = Math.max(maxT, T.abs());
        }

        System.out.println("max |S(jw)| = " + maxS);
        System.out.println("max |T(jw)| = " + maxT);
        // These can be compared with target bounds implied by weights W_P, W_N, W_D.
    }
}

File: Linear_Control_Course/Chapter24/Lesson2/Chapter24_Lesson2.cpp
#include <iostream>
#include <complex>
#include <vector>
#include <cmath>

using cd = std::complex<double>;

// Evaluate P(jw) = 1 / (jw^2 + 2*zeta*wn*jw + wn^2)
cd Pjw(double w, double zeta, double wn) {
    cd jw(0.0, w);
    cd denom = jw * jw + cd(2.0 * zeta * wn, 0.0) * jw + cd(wn * wn, 0.0);
    return cd(1.0, 0.0) / denom;
}

// Controller C(s) = K (s + z) / s evaluated at s = jw
cd Cjw(double w, double K, double z) {
    cd jw(0.0, w);
    return K * (jw + cd(z, 0.0)) / jw;
}

// First-order performance weight W_P(s) = (s/M + wp) / (s + wp*A)
cd WPjw(double w, double M, double A, double wp) {
    cd jw(0.0, w);
    return (jw / M + cd(wp, 0.0)) / (jw + cd(wp * A, 0.0));
}

// Multiplicative uncertainty weight W_D(s) = (s/alpha + wD) / (s + wD)
cd WDjw(double w, double alpha, double wD) {
    cd jw(0.0, w);
    return (jw / alpha + cd(wD, 0.0)) / (jw + cd(wD, 0.0));
}

int main() {
    double zeta = 0.7, wn = 10.0;
    double K = 20.0, z = 2.0;
    double M_P = 1.5, A_P = 1e-3, wP = 1.0;
    double alpha = 1.5, wD = 30.0;

    std::vector<double> wgrid;
    for (int i = 0; i <= 500; ++i) {
        double frac = static_cast<double>(i) / 500.0;
        double w = std::pow(10.0, -2.0 + 5.0 * frac); // 1e-2 to 1e3
        wgrid.push_back(w);
    }

    double max_WP_S = 0.0;
    double max_WD_T = 0.0;

    for (double w : wgrid) {
        cd P = Pjw(w, zeta, wn);
        cd C = Cjw(w, K, z);
        cd L = C * P;
        cd S = cd(1.0, 0.0) / (cd(1.0, 0.0) + L);
        cd T = L / (cd(1.0, 0.0) + L);

        cd WP = WPjw(w, M_P, A_P, wP);
        cd WD = WDjw(w, alpha, wD);

        double WP_S = std::abs(WP * S);
        double WD_T = std::abs(WD * T);

        if (WP_S > max_WP_S) max_WP_S = WP_S;
        if (WD_T > max_WD_T) max_WD_T = WD_T;
    }

    std::cout << "max |W_P(jw) S(jw)| = " << max_WP_S << std::endl;
    std::cout << "max |W_D(jw) T(jw)| = " << max_WD_T << std::endl;
    return 0;
}
